### PR TITLE
Let kill query bypass the resource group queues

### PR DESCRIPTION
**Pull request by ShashwatArghode**

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java [hunk 1]
```diff
@@ -270,6 +271,14 @@ public QueryInfo getFullQueryInfo()
                 .orElseGet(() -> stateMachine.updateQueryInfo(Optional.empty()));
     }
 
+    @Override
+    public Optional<Statement> getStatement()
+    {
+        return tryGetQueryExecution()
+                .map(QueryExecution::getStatement)
+                .orElseGet(() -> Optional.empty());
```
Comment (Praveen2112) [inline core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java:279]: I guess this is not required. `Optional#map` returns `Optional.empty if the element is not present.

File: core/trino-main/src/main/java/io/trino/execution/resourcegroups/InternalResourceGroup.java [hunk 2]
```diff
@@ -586,6 +589,17 @@ public InternalResourceGroup getOrCreateSubGroup(String name)
...
 
+    public boolean isKillQuery(Optional<Statement> statement)
+    {
+        if (statement.isPresent() && statement.get() instanceof Call) {
+            Call call = (Call) statement.get();
+            if (call.getName().toString().endsWith(KILL_QUERY)) {
+                return true;
+            }
+        }
+        return false;
```
Comment (Praveen2112) [inline core/trino-main/src/main/java/io/trino/execution/resourcegroups/InternalResourceGroup.java:600]:
```
        statement.filter(call -> call instanceof Call)
                .map(Call.class::cast)
                .map(Call::getName)
                .filter(name -> name.getSuffix().equals(KILL_QUERY));
```
                
Can we something like this

File: core/trino-main/src/main/java/io/trino/execution/resourcegroups/InternalResourceGroup.java [hunk 3]
```diff
@@ -596,14 +610,17 @@ public void run(ManagedQueryExecution query)
             InternalResourceGroup group = this;
             boolean canQueue = true;
             boolean canRun = true;
-            while (true) {
-                canQueue = canQueue && group.canQueueMore();
-                canRun = canRun && group.canRunMore();
-                if (group.parent.isEmpty()) {
-                    break;
+            if (!isKillQuery(query.getStatement())) {
```
Comment (Praveen2112) [inline core/trino-main/src/main/java/io/trino/execution/resourcegroups/InternalResourceGroup.java:613]: Can we merge this with the `while` loop condition ?

**Review by Praveen2112 — COMMENTED** [review]
Thanks for working on this. 
It would be better if we could have some tests for it. Maybe to run a big query in a very small resource group of size 1 and check if we are able to kill that query.

**Review by phd3 — COMMENTED** [review]
Thanks @ShashwatArghode! Since UI/restcalls allow effectively doing the same operations without resourcing constraint, this behavior looks consistent with it. 

However, I think we can also implement this by adding a default root resource group in `InternalResourceGroupManager`, which is matched before others, and tackles all "kill" queries. That way, we can still have _some_ configurable upper bound on running/queued queries to help us avoid someone accidentally sending way too many kill queries. @Praveen2112 @dain wdyt?

File: core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java (hunk 1 above)
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java:279]: `.map(QueryExecution::getStatement)` would need to become `flatMap` (as `QueryExecution::getStatement` returns an optional)

File: core/trino-main/src/main/java/io/trino/execution/QueryExecution.java [hunk 4]
```diff
@@ -74,6 +76,11 @@
 
     boolean shouldWaitForMinWorkers();
 
+    default Optional<Statement> getStatement()
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/QueryExecution.java:79]: I am not convinced we want to expose AST-level information here. cc @martint

File: core/trino-main/src/main/java/io/trino/execution/resourcegroups/InternalResourceGroup.java [hunk 5]
```diff
@@ -586,6 +589,17 @@ public InternalResourceGroup getOrCreateSubGroup(String name)
         }
     }
 
+    public boolean isKillQuery(Optional<Statement> statement)
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/resourcegroups/InternalResourceGroup.java:592]: It effectively archives the goal, but also introduces tight coupling between previously unrelated parts of the system (resource groups system, and one of the way to kill a query). We should try to identify more higher level abstraction that would address the use-case.

**Comment by findepi** [comment]
@phd3 i think the problem may be more general that just kill queries.
For example, i don't want to allow a user to execute two "big" queries at a time, but 
i don't want to prevent them from running a DESCRIBE, so that they can work on their
second query while the first one is executing.

@ShashwatArghode what is the use-case you're trying to solve and what currently prevents resource groups from from meeting your requirements?

**Comment by phd3** [comment]
> For example, i don't want to allow a user to execute two "big" queries at a time, but
i don't want to prevent them from running a DESCRIBE, so that they can work on their
second query while the first one is executing.

For such cases, wouldn't it be okay leaving it up to the admin to setup different resource groups using query types? I think the kill_query case is a bit more interesting. It's not supposed to do any metadata/data processing and performs an operation available through other interfaces (UI / rest). But resource groups can put the user in kind of a deadlock, if their query is taking super long to execute, and now they want to run a different query. But they can't even kill that original query through SQL. So a separate default resource group (for killing) with sensible upper bound on running/queued queries may be useful.  

As an alternate approach, may be we can look into making the queryType more granular, or nested - e.g. identifying kill_query as a query type and let admins handle it.

**Comment by colebow** [comment]
:wave: @ShashwatArghode - this PR is inactive and doesn't seem to be under development. If you'd like to continue work on this at any point in the future, feel free to re-open.

**Pull request closed without merging**